lib/os_io.c

```diff
@@ -37,6 +37,130 @@ ___io_module ___io_mod =
 
 /*---------------------------------------------------------------------------*/
 
+/* poll dynamic fdset memory management. */
+
+#ifdef USE_poll
+#define ___fdset_state_size(ps)     ps->os.fdset.size
+#define ___fdset_state_readfds(ps)  ps->os.fdset.readfds
+#define ___fdset_state_writefds(ps) ps->os.fdset.writefds
+
+static int ___fdset_heap_overflow;
+
+static int ___fdset_state_init (___processor_state ps)
+{
+  void *readfds = NULL;
+  void *writefds = NULL;
+
+  readfds  = ___ALLOC_MEM (MAX_CONDVARS/8);
+  if (!readfds)
+    goto error;
+
+  writefds = ___ALLOC_MEM (MAX_CONDVARS/8);
+  if (!writefds)
+    goto error;
+
+  memset (readfds, 0, MAX_CONDVARS/8);
+  memset (writefds, 0, MAX_CONDVARS/8);
+  ___fdset_state_readfds (ps) = readfds;
+  ___fdset_state_writefds (ps) = writefds;
+  ___fdset_state_size (ps) = MAX_CONDVARS;
+
+  return 0;
+
+ error:
```

Comment:
Ok, will use ___FREE_MEM instead of ___free_mem; didn't know about it.

Instead of changing the straightline implementation of the code, I can put a check on whether the pointers are not NULL in the error handler. That way the nesting remains small and the code is possibly easier to read.

But I can also change to the nested if style if you feel strongly about it.